### PR TITLE
Lambda Layer

### PR DESCRIPTION
**Pull request by farizrahman4u**
Lambda layer can be used to implement any arbitrary function layer.

### CODE REVIEW

**Comment by farizrahman4u** [comment]
**Key features:**

-Serializable
-Any layer can be could be modeled by adding a ```Lambda``` layer over a ```join``` ```Merge``` layer
-The ```output_shape``` could be a ```tuple```, or a function of the previous layer.

**Comment by EderSantana** [comment]
Relevant issue https://github.com/fchollet/keras/issues/883. Could you add some tests to a basic operation and serialization? Also, I suggest you to call this layer `MaskedLambda` and we may need `Lambda` to inherit from `Layer` instead. If you try `MaskedLambda` after, say dense, Keras may complain about it.

**Comment by farizrahman4u** [comment]
@EderSantana  Not yet! We have to figure out a way to avoid deserializing (```marshal.loads```) the function every time ```get_output()``` is called. This could be serious performance bottle neck, or maybe not. Only after conducting a performance test  ( by implementing an existing layer using Lambda layer and comparing train and test duration of 2 models having each of the original layer and lambda version) we can know for sure.

**Comment by farizrahman4u** [comment]
@EderSantana  Test for Lambda layer added.

**Comment by farizrahman4u** [comment]
Done! Now works on both 2.7 and 3.4

**Comment by farizrahman4u** [comment]
Added Serialization test. Works like a charm on 2.7.  3.4 upset again

**Comment by farizrahman4u** [comment]
And we are back! Test now works on both python 2.7 and 3

**Comment by farizrahman4u** [comment]
@fchollet  What do you think?

File: keras/layers/core.py [hunk 1]
```diff
@@ -345,7 +346,8 @@ def get_output(self, train=False):
             for i in range(len(self.layers)):
                 X = self.layers[i].get_output(train)
                 if X.name is None:
-                    raise ValueError("merge_mode='join' only works with named inputs")
+                    inputs['input' + str(i)] = self.layers[i].get_output(train)
```
Comment (fchollet) [inline keras/layers/core.py]: Too implicit. This should only work when layers are explicitly named.

**Comment by fchollet** [comment]
> @fchollet What do you think?

I think marshaling is a good solution to the serialization issue. This is looking good.

I haven't checked out the tests yet. As for the layer itself, what it needs would be:
- a docstring specifying the meaning and format of the arguments. Much like [this](https://github.com/fchollet/keras/blob/master/keras/layers/convolutional.py#L409). It is not necessary for most layers, but in this case it is not immediately obvious what the purpose of the Lambda layer is, and because of polymorphism it is not obvious either how to use it. So that should be made clear.
- enforcement of argument expectations via type checks and exception raising. Again, necessary in this case because of the polymorphism.

**Comment by farizrahman4u** [comment]
@fchollet Please help me out. Why is ```join``` ```Merge``` returning an ```OrderedDic``` rather than the more intuitive ```list``` of ```tensor```? What are we accomplishing by making sure that the layer outputs are named?

**Comment by fchollet** [comment]
> What are we accomplishing by making sure that the layer outputs are named (explicitly) ?

Two things:
- more flexible layer connectivity. Let's say a layer returns a list of tensors [A, B, C] and you want to connect to a layer that takes 3 inputs, but will expect the order [C, A, B]. It's easy and explicit to achieve if you are returning dictionaries rather than lists, and if you can pass a connectivity map that says "connect A to input 1, C to input 0 and B to input 2". 
- API consistency. To achieve recursive compositionality, which is a big strength of Keras, you need layers to behave like containers (Graph / Sequential). The Graph layer takes as inputs dictionaries of tensors and returns the same. So we need multi-io layers to follow the same API and to return dictionaries of tensors.

As for implicit naming -- in general we should no act in "creative" ways that aren't explicitly controlled by the user. Picking names for layers (that won't be displayed anywhere) is too much interference on our part, at least with the current API. We can consider it when we set up a global layer naming scheme, where every layer will take a `name` argument that will be filled intelligently by default. Such a scheme might come roughly at the same time as the future abstract backend.

**Comment by farizrahman4u** [comment]
@fchollet  Agreed. But how would I set names for the inputs? I tried the ```set_name()``` of the models being merged, but doesnt seem to work.

Also, implicit naming of layers using a global nomenclature seems to be a great idea. A layer name could be Layer type + index. Like fc1, fc2, relu1, conv1 etc.

**Comment by farizrahman4u** [comment]
Need some help here.. How to set the names for join merge? Can find any implementation of join merge anywhere.

**Comment by farizrahman4u** [comment]
Need some help here. How to set the names for layers being merged? Cant find any implementation of join merge, not even a test.

**Comment by EderSantana** [comment]
You do that with a graph
```python
model.add(layer, inputs=['prev_1', 'prev_2'], merge_mode='join')
```

Keras sets the names of the outputs `prev_1`and `prev_2`. What exactly is your problem?

**Comment by farizrahman4u** [comment]
@EderSantana  My exact problem is that I want to make the 'Sum merge using Lambda layer' example given in the first comment work, without making any changes to the ```Merge``` class.

**Comment by farizrahman4u** [comment]
@EderSantana  My exact problem is that I want to make the 'Sum merge using Lambda layer' example given in the first comment work, without making any changes to the ```Merge``` class.

**Comment by EderSantana** [comment]
I understand now, just make sure the `Tensors` you are summing up have the `name` property. For example
```python
X = T.matrix('any_name_works')  # or
X = T.matrix()
X.name = 'any_name_at_all
```

**Comment by farizrahman4u** [comment]
@EderSantana  But the example given above is a sequential layer. The merge layer gets inputs from the left and right models, whose outputs do not have any name

**Comment by farizrahman4u** [comment]
@EderSantana  Then we would have to use Graph for a purely Sequential problem. Let me try if I can do something with least changes to the working of Merge layer.

File: keras/layers/core.py [hunk 2]
```diff
@@ -276,7 +277,7 @@ def get_config(self):
 
 
 class Merge(Layer):
-    def __init__(self, layers, mode='sum', concat_axis=-1):
+    def __init__(self, layers, mode='sum', concat_axis=-1, auto_name=False):
```
Comment (farizrahman4u) [inline keras/layers/core.py]: @EderSantana  This should do the job

**Comment by EderSantana** [comment]
If are merging two branches how is that Sequential? I didn't get the problem them...

**Comment by farizrahman4u** [comment]
@EderSantana Merge layer is often used in Sequential models. Checkout the Merge example : http://keras.io/layers/core/

**Comment by EderSantana** [comment]
isn't there a way to name the outputs right before merge?

**Comment by farizrahman4u** [comment]
@EderSantana No, I couldnt find.  I thought ```left.set_name('left')``` would work, but it doesnt.

**Comment by EderSantana** [comment]
here is how `container.Graph` does it https://github.com/fchollet/keras/blob/master/keras/layers/containers.py#L236

**Comment by farizrahman4u** [comment]
@EderSantana  But Francois told me not to implicitly name the tensors, without the knowledge of the user.

**Comment by EderSantana** [comment]
But this is for the test right? You can always document and tell the user how to do it their way.

**Comment by farizrahman4u** [comment]
* The ```join``` merge was written with ```Graph``` in mind, not ```Sequential``` 
* Lambda is the only layer capable of sitting on top of a ```join``` ```Merge``` in a ```Sequential``` model , so adding an argument to ```Merge`` which will be used only when a ```Lambda``` is on top is not a good idea.
* If Lambda can not be stacked over a ```join``` ```Merge```, its worthless.  ```Lambda``` would be just a  custom activation layer.

Above facts suggest that we need a ```LambdaMerge``` layer. So that we can leave ```join Merge``` alone.

**Comment by EderSantana** [comment]
IMHO the real problem is the fact that Sequential does not implement "naming" yet. This shouldn't be the way things work. Naming is important.

**Comment by fchollet** [comment]
> IMHO the real problem is the fact that Sequential does not implement "naming" yet.

Yes, this is the real issue here. As mentioned previously, I hope we will soon have a framework-wide naming system, where every layer / container will accept a `name` argument and will have a `set_name` method, and where names not provided will be automatically generated using a scheme that will be both intuitive and explicitly documented.

We had a PR for a pretty good naming system some time ago, but it made use of initialization functions as name-injection mechanism, which we decided against. As we introduce an abstract backend, we will have access to a variable instantiation system that will replace `floatX` and the like, and that will be used as the go-to name-injection mechanism.

File: keras/layers/core.py [hunk 3]
```diff
@@ -899,3 +903,177 @@ def get_config(self):
...
+    -----------
+    output_shape of previous layer
+
+    Output shape
+    ------------
+    Specified by output_shape argument
+
+    Arguments
+    ---------
+    function - The function to be evaluated. Takes one argument : ouput of previous layer
```
Comment (fchollet) [inline keras/layers/core.py]: Typo: output. It should be clarified that it needs to be a valid Theano expression.

File: keras/layers/core.py [hunk 4]
```diff
@@ -899,3 +903,177 @@ def get_config(self):
...
+    output_shape - Expected output shape from function. Could be a tuple or a function of input layer
```
Comment (fchollet) [inline keras/layers/core.py]: Function of input layer *shape*.

**Comment by fchollet** [comment]
I think `LambdaMerge` is pretty good solution for what you are trying to do, in the current state of things. It's not entirely clear if anybody would actually use it, though. I understand the use case, but it seems somewhat arcane.

**Comment by farizrahman4u** [comment]
@fchollet What do you think about the LambdaMerge layer? It inherently solves the naming issue, as ```join``` ```Merge``` is not involved.

**Comment by farizrahman4u** [comment]
@fchollet  If there is no ```LambdaMerge``` we will fill up the ```Merge``` class with too many modes, and each of these modes will be used even rarely.

**Comment by farizrahman4u** [comment]
@fchollet  Also, the Lambda layer kind of makes Keras 'Turing complete' in a sense.

**Comment by fchollet** [comment]
> If there is no LambdaMerge we will fill up the Merge class with too many modes, and each of these modes will be used even rarely.

I would agree with this point. I see the use case, and I think it's a good solution. It does contribute to make Keras more flexible.

Then again of course, it does makes for a much simpler UX to have the most common merge modes covered in the base `Merge` layers.

By the way, this solution works great in terms of compatibility with single-output / single-input layers (unnamed), but as far as I can tell it isn't compatible with multi-io layers (i.e. `Graph`, with named inputs / ouputs).

**Comment by farizrahman4u** [comment]
> but as far as I can tell it isn't compatible with multi-io layers (i.e. Graph, with named inputs / ouputs).
I havent checked out graphs yet. I assume its like a mesh network, each node with multiple inputs and outputs.
The ```Lambda``` layer inherits from ```Layer```, so it could be used anywhere a `Dense` layer is used.
The `LambdaMerge` layer is exclusive for `Sequential`. In a `Graph`, similar functionality could be obtained by first merging using `join` mode, and then a `Lambda` layer, just like any other layer you would add to a `Graph` after `join` merge. Again, this might sound stupid if my assumption about the working of `Graph` is wrong.

File: keras/layers/core.py (hunk 4 above)
Comment (farizrahman4u) [inline keras/layers/core.py]: Done

**Comment by farizrahman4u** [comment]
Apparently, there is some problem with serialization in python 2.7. Python 3 is good

**Comment by farizrahman4u** [comment]
@fchollet  OK, we are good. What next?

**Comment by fchollet** [comment]
Let me review this in detail, then we will merge. It might be a few days until I find the time.

**Comment by farizrahman4u** [comment]
@fchollet  OK, cool.

File: keras/layers/core.py [hunk 5]
```diff
@@ -899,3 +903,182 @@ def get_config(self):
...
+    ------------
+    Specified by output_shape argument
+
+    Arguments
+    ---------
+    function - The function to be evaluated. Takes one argument : output of previous layer
+    output_shape - Expected output shape from function. Could be a tuple or a function of the shape of the input
+    """
+
+    def __init__(self, function, output_shape=None, ndim=2):
```
Comment (fchollet) [inline keras/layers/core.py]: What is the justification for the `ndim` argument here? The input dimension is managed via `set_input`, and it is automated as part of the shape inference system.

File: keras/layers/core.py [hunk 6]
```diff
@@ -899,3 +903,182 @@ def get_config(self):
...
+                self._output_shape = marshal.dumps(output_shape.__code__)
+            else:
+                self._output_shape = marshal.dumps(output_shape.func_code)
+
+    @property
+    def output_shape(self):
+        if self._ouput_shape is None:
+            return self.input_shape
+        elif type(self._output_shape) == tuple:
+            return self._output_shape
```
Comment (fchollet) [inline keras/layers/core.py]: For the `output_shape` argument to be consistent with the `input_shape` constructor argument, it should not include the first dimension (the samples dimension). As such, the return value here would be `(self.input_shape[0],) + self._output_shape`.

This behavior should also be documented.

File: keras/layers/core.py [hunk 7]
```diff
@@ -899,3 +903,182 @@ def get_config(self):
...
+            output_shape_func = marshal.loads(self._output_shape)
+            output_shape_func = types.FunctionType(output_shape_func, globals())
+            shape = output_shape_func(self.previous.output_shape)
+            if type(shape) not in {list, tuple}:
+                raise Exception("output_shape function must return a tuple")
+            return tuple(shape)
+
+    def get_output(self, train=False):
+        func = marshal.loads(self.function)
+        func = types.FunctionType(func, globals())
```
Comment (fchollet) [inline keras/layers/core.py]: To be honest, this feels very much unsafe.

File: keras/layers/core.py [hunk 8]
```diff
@@ -899,3 +903,182 @@ def get_config(self):
...
+                self._output_shape = marshal.dumps(output_shape.__code__)
+            else:
+                self._output_shape = marshal.dumps(output_shape.func_code)
+
+    @property
+    def output_shape(self):
+        if self._output_shape is None:
+            return self.layers[0].input_shape
+        elif type(self._output_shape) == tuple:
+            return self._output_shape
```
Comment (fchollet) [inline keras/layers/core.py]: Same remark as before regarding `output_shape`.

File: tests/auto/test_sequential_model.py [hunk 9]
```diff
@@ -297,7 +377,7 @@ def test_count_params(self):
         model.compile('sgd', 'binary_crossentropy')
 
         self.assertEqual(n, model.count_params())
-
+
```
Comment (fchollet) [inline tests/auto/test_sequential_model.py]: Unnecessary whitespace.

File: tests/auto/test_sequential_model.py [hunk 10]
```diff
@@ -275,7 +278,84 @@ def test_merge_overlap(self):
...
+        model.add(Lambda(activation))
+        model.load_weights('temp.h5')
+        model.compile(loss='categorical_crossentropy', optimizer='rmsprop')
+
+        nloss = model.evaluate([X_train, X_train], y_train, verbose=0)
+        print(nloss)
+        assert(loss == nloss)
+        
+        print ('test serializing')
+        func = None # Make sure that the model has the function code, not just the function name.
```
Comment (fchollet) [inline tests/auto/test_sequential_model.py]: Wouldn't `del func` make more sense here?

**Comment by fchollet** [comment]
I think this is looking good, despite the general unsafeness of using marshal / FunctionType. Main remark is to make the `output_shape` argument consistent with `input_shape` (i.e. don't include the samples dimension). And input specification via `ndim` should be removed.

**Comment by fchollet** [comment]
One more thing: as per e.g. [this comment](https://groups.google.com/forum/#!topic/keras-users/SzQeBgXA70k), and following contributor guidelines, these changes should be documented.

**Comment by farizrahman4u** [comment]
@fchollet  

*  Everything other than documentation has been done. Need some time on that.
* Regarding safety: is there some way to make sure a function is read only in python? Like a const function in C++? Anyway, Lambda layer is for advanced users, so this should not be a big issue I think.

**Comment by fchollet** [comment]
> Lambda layer is for advanced users, so this should not be a big issue I think.

Agreed. I don't think it's a big problem for us.

**Comment by farizrahman4u** [comment]
@fchollet  I need confirmation on the current API of Lambda and LambdaMerge layers. I am about to start the documentation.

**Comment by fchollet** [comment]
The current API is fine IMO. We can go with it.

**Comment by farizrahman4u** [comment]
Ok. Doing the docs now.

**Comment by farizrahman4u** [comment]
@fchollet Please go through the documentation.

**Comment by fchollet** [comment]
Great, thanks! I fixed the coding in the documentation, and merged.

**Comment by farizrahman4u** [comment]
@fchollet Update keras.io too.
